diff --git a/app/src/main/java/com/example/ergometerapp/MainActivity.kt b/app/src/main/java/com/example/ergometerapp/MainActivity.kt
index 38ccb69..3e3ec65 100644
--- a/app/src/main/java/com/example/ergometerapp/MainActivity.kt
+++ b/app/src/main/java/com/example/ergometerapp/MainActivity.kt
@@ -79,6 +79,8 @@ class MainActivity : ComponentActivity() {
         registerForActivityResult(ActivityResultContracts.RequestPermission()) { granted ->
             if (granted) {
                 Log.d("BLE", "BLUETOOTH_CONNECT granted")
+                bleClient.connect("E0:DF:01:46:14:2F")
+                hrClient.connect("24:AC:AC:04:12:79")
             } else {
                 Log.d("BLE", "BLUETOOTH_CONNECT denied")
             }
@@ -210,7 +212,6 @@ class MainActivity : ComponentActivity() {
                 }
             }
         }
-        ensureBluetoothPermission()

         bleClient = FtmsBleClient(
             context = this,
@@ -250,8 +251,6 @@ class MainActivity : ComponentActivity() {
         }

         // TODO: Move hardcoded MACs to configuration or device discovery flow.
-        bleClient.connect("E0:DF:01:46:14:2F")
-
         hrClient = HrBleClient(this)
         { bpm ->
             heartRateState.value = bpm
@@ -259,7 +258,7 @@ class MainActivity : ComponentActivity() {

         }

-        hrClient.connect("24:AC:AC:04:12:79")
+        ensureBluetoothPermission()
     }
         /**
      * Keeps the screen awake during an active session to avoid lost telemetry
@@ -336,6 +335,9 @@ class MainActivity : ComponentActivity() {
             != PackageManager.PERMISSION_GRANTED
         ) {
             requestBluetoothConnectPermission.launch(Manifest.permission.BLUETOOTH_CONNECT)
+        } else {
+            bleClient.connect("E0:DF:01:46:14:2F")
+            hrClient.connect("24:AC:AC:04:12:79")
         }
     }

